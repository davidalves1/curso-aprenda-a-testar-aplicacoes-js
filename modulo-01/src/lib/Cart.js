import find from 'lodash/find';
import lshRemove from 'lodash/remove';

export default class Cart {
  items = [];

  add(item) {
    const { product } = item;

    if (find(this.items, { product })) {
      lshRemove(this.items, { product });
    }

    this.items.push(item)
  }

  remove(product) {
    lshRemove(this.items, { product });
  }

  getTotal() {
    return this.items.reduce((acc, curr) => {
      return acc + curr.quantity * curr.product.price;
    }, 0);
  }
}
