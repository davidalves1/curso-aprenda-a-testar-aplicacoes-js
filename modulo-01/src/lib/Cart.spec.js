import Cart from './Cart';

describe('Cart', () => {
  let cart;

  let product = {
    title: 'Adidas running shoes - men',
    price: 19990,
  };

  let product2 = {
    title: 'Adidas running shoes - women',
    price: 18990,
  };

  beforeEach(() => {
    cart = new Cart();
  });

  describe('getTotal()', () => {
    it('should return 0 on getTotal if cart is empty', () => {
      expect(cart.getTotal().getAmount()).toEqual(0);
    });

    it('should multiply quantity and price to receive the total amount', () => {
      const item = {
        quantity: 2,
        product,
      };

      cart.add(item);

      expect(cart.getTotal().getAmount()).toEqual(39980);
    });

    it('should ensure no more than one product exists at a time', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product,
        quantity: 1,
      });

      expect(cart.getTotal().getAmount()).toEqual(19990);
    });

    it('should update total when a product gets included and then removed', () => {
      cart.add({
        product,
        quantity: 2,
      });

      cart.add({
        product: product2,
        quantity: 1,
      });

      cart.remove(product)

      expect(cart.getTotal().getAmount()).toEqual(18990);
    });
  });

  describe('checkout', () => {
    it('should return an object with the total and the list of items when call sumary', () => {
      cart.add({
        product,
        quantity: 3,
      });

      cart.add({
        product: product2,
        quantity: 2,
      });

      expect(cart.sumary()).toMatchSnapshot();
      expect(cart.getTotal().getAmount()).toBeGreaterThan(0);
    });

    it('should return an object with the total and the list of items', () => {
      cart.add({
        product,
        quantity: 3,
      });

      cart.add({
        product: product2,
        quantity: 2,
      });

      expect(cart.checkout()).toMatchSnapshot();
    });

    it('should reset the cart when checkout() is called', () => {
      cart.add({
        product,
        quantity: 3,
      });

      cart.checkout()

      expect(cart.getTotal().getAmount()).toEqual(0);
    });
  });

  describe('special discounts', () => {
    it('should apply percentage discount quantity above minimum is passed', () => {
      const condition = {
        percentage: 30,
        minimum: 2
      };

      cart.add({
        product,
        condition,
        quantity: 3,
      });

      expect(cart.getTotal().getAmount()).toEqual(41979)
    });

    it('should NOT apply percentage discount quantity equal or below minimum', () => {
      const condition = {
        percentage: 30,
        minimum: 2
      };

      cart.add({
        product,
        condition,
        quantity: 2,
      });

      expect(cart.getTotal().getAmount()).toEqual(39980)
    });

    it('should apply quantity discount for even quantities', () => {
      const condition = {
        quantity: 2,
      };

      cart.add({
        product,
        condition,
        quantity: 4,
      });

      expect(cart.getTotal().getAmount()).toEqual(39980)
    });

    it('should apply quantity discount for even quantities', () => {
      const condition = {
        quantity: 2,
      };

      cart.add({
        product,
        condition,
        quantity: 1,
      });

      expect(cart.getTotal().getAmount()).toEqual(19990)
    });

    it('should apply quantity discount for odd quantities', () => {
      const condition = {
        quantity: 2,
      };

      cart.add({
        product,
        condition,
        quantity: 3,
      });

      expect(cart.getTotal().getAmount()).toEqual(35982)
    });
  });
});
