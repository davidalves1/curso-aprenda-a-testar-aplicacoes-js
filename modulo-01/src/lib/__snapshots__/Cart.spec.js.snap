// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout should return an object with the total and the list of items 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 19990,
        "title": "Adidas running shoes - men",
      },
      "quantity": 3,
    },
    Object {
      "product": Object {
        "price": 18990,
        "title": "Adidas running shoes - women",
      },
      "quantity": 2,
    },
  ],
  "total": 97950,
}
`;

exports[`Cart checkout should return an object with the total and the list of items when call sumary 1`] = `
Object {
  "items": Array [
    Object {
      "product": Object {
        "price": 19990,
        "title": "Adidas running shoes - men",
      },
      "quantity": 3,
    },
    Object {
      "product": Object {
        "price": 18990,
        "title": "Adidas running shoes - women",
      },
      "quantity": 2,
    },
  ],
  "total": 97950,
}
`;
